versioning/src/main/scala/org/typelevel/sbt/TypelevelVersioningPlugin.scala

```diff
@@ -0,0 +1,82 @@
+package org.typelevel.sbt
+
+import sbt._, Keys._
+import com.typesafe.sbt.GitPlugin
+import com.typesafe.sbt.SbtGit.git
+import org.typelevel.sbt.kernel.V
+
+import scala.util.Try
+import org.typelevel.sbt.kernel.GitHelper
+
+object TypelevelVersioningPlugin extends AutoPlugin {
+
+  override def requires = GitPlugin
+  override def trigger = allRequirements
+
+  object autoImport {
+    lazy val tlBaseVersion =
+      settingKey[String]("The base version for the series your project is in. e.g., 0.2, 3.5")
+    lazy val tlHashSnapshots =
+      settingKey[Boolean]("If true, a hash version implies this is a snapshot (default: true).")
+  }
+
+  import autoImport._
+
+  override def buildSettings: Seq[Setting[_]] = Seq(
+    versionScheme := Some("early-semver"),
+    tlHashSnapshots := true,
```

Comment:
Is this going to add hash release to Maven Central by default?  I know mutable snapshots are problematic, but those non-monotonic releases cluttering Maven Central are gross in their own right.  I'd rather make it easy for people to cut milestones.

Reply:
Nope, the default is that snapshots go to the snapshots repo. However setting this `tlHashSnapshots` to false would achieve stable snapshots to Maven Central.

Btw, the versioning achieves monotonicity by also including the distance from the last non-prerelease tag in the same series, if available, before the hash.

Reply:
Btw, this plugin published itself as a snapshot:
https://s01.oss.sonatype.org/content/repositories/snapshots/com/armanbilge/sbt-typelevel-ci-release_2.12_1.0/0.4-a57db49-SNAPSHOT/

I temporarily switched it to publish to Maven Central to test the signing machinery.

Reply:
I think a lot of confusion comes from imprecise vocab.  I think of "snapshots" as anything that has `SNAPSHOT` in the version and thus is treated as a mutable artifact and banished by common policy to the `snapshots` directory of the repository.  And a "release" is something that does not have `SNAPSHOT` in the name, but may be a "hash release" by having a hash in what you call the `V.prelease`.

So I think your setup makes sense, and the default does not offend me, but there's enough confusion around the name that we should document what happens when it's true and false.

Reply:
💯 I completely agree.

To add to the confusion, this decision is made at versioning time (hence, why it is in the versioning plugin). I.e. whether you append `-SNAPSHOT` to your version number. That then determines where it gets published.

When this setting is `true`, any version with a hash in it is deemed a `-SNAPSHOT`.

When this setting is `false`, `-SNAPSHOT` is appended to the version number _only_ if the git repo has uncommitted changes in it. Which should never happen in CI.

👍 to good docs with examples

Reply:
Renamed to `tlUntaggedAreSnapshots`, see below.